src/core/UbqScaledImageGenerator.cpp

```diff
@@ -0,0 +1,155 @@
+    switch (result) {
+        case SkCodec::kSuccess:
+        case SkCodec::kIncompleteInput:
+        case SkCodec::kErrorInInput:
+            return true;
+        default:
+            return false;
+    }
+}
+
+bool UbqScaledImageGenerator::getPixels(const SkImageInfo& info,
+                                        void* pixels,
+                                        size_t rowBytes,
+                                        const SkCodec::Options* options) {
+    SkPixmap dst(info, pixels, rowBytes);
+
+    std::function<bool(const SkPixmap&)> decode;
+    if (!needsScaling()) {  // Image is returned "as is", no fallback
+        decode = [this, options](const SkPixmap& pm) {
+            SkCodec::Result result = fCodec->getPixels(pm, options);
+            return isValidDecode(result);
+        };
+    } else {  // Query scaled version
+        decode = [this, options, info](const SkPixmap& pm) {
+            // Ask codec to return image at requested size
+            SkCodec::Result getPixelsResult = fCodec->getPixels(pm, options);
+            if (isValidDecode(getPixelsResult)) {
+                return true;
+            } else {
+                // Fall back to decode at closest dimension followed by forced downscale
+                float limit = fMaxImageSize;
+                float width = fCodec->getInfo().width();
+                float height = fCodec->getInfo().height();
+                auto scale = std::min(limit / height, limit / width);
+
+                // Ask codec for supported size, that best matches the target scale
+                SkISize decodableSize = fCodec->getScaledDimensions(scale);
+                auto [image, getImageResult] = fCodec->getImage(info.makeDimensions(decodableSize));
+
+                // Downscale decoded image into dstMap
+                if (isValidDecode(getImageResult)) {
+                    return image->scalePixels(pm, SkSamplingOptions(SkFilterMode::kLinear));
+                }
+
+                return false;
+            }
+        };
+    }
+
+    // Account for EXIF
+    return SkPixmapPriv::Orient(dst, fCodec->getOrigin(), decode);
+}
+
+bool UbqScaledImageGenerator::onGetPixels(const SkImageInfo& requestInfo,
+                                          void* requestPixels,
+                                          size_t requestRowBytes,
+                                          const Options& options) {
+    return this->getPixels(requestInfo, requestPixels, requestRowBytes, nullptr);
+}
+
+bool UbqScaledImageGenerator::onQueryYUVAInfo(
+        const SkYUVAPixmapInfo::SupportedDataTypes& supportedDataTypes,
+        SkYUVAPixmapInfo* yuvaPixmapInfo) const {
+    if (!needsScaling()) {
+        return fCodec->queryYUVAInfo(supportedDataTypes, yuvaPixmapInfo);
+    }
+
+    return false;
+}
+
+bool UbqScaledImageGenerator::onGetYUVAPlanes(const SkYUVAPixmaps& yuvaPixmaps) {
+    if (!needsScaling()) {
+        switch (fCodec->getYUVAPlanes(yuvaPixmaps)) {
```

Comment:
not that it matters, but I think you can use `return isValidDecode(fCodec->getYUVAPlanes(yuvaPixmaps));` here as well